### PR TITLE
feat: support IVF_FLAT, binary vectors and hamming distance

### PR DESCRIPTION
**Pull request by BubbleCal**
binary vectors and hamming distance can work on only IVF_FLAT, so introduce them all in this PR.

### CODE REVIEW

File: docs/src/search.md [hunk 1]
```diff
@@ -107,6 +111,33 @@ an ANN search means that using an index often involves a trade-off between recal
 See the [IVF_PQ index](./concepts/index_ivfpq.md) for a deeper description of how `IVF_PQ`
 indexes work in LanceDB.
 
+## Binary vector
+
+LanceDB supports binary vectors as a data type, and has the ability to search binary vectors with hamming distance. The binary vectors are stored as uint8 arrays:
```
Comment (wjones127) [inline docs/src/search.md:116]: Is each bit in the uint8 array a value? or each byte? Could you describe that format a little more?

File: docs/src/search.md [hunk 2]
```diff
@@ -107,6 +111,33 @@ an ANN search means that using an index often involves a trade-off between recal
...
+            "vector": np.random.randint(0, 256, size=128),
+        }
+        for i in range(1024)
+    ]
+
+    tbl = db.create_table("my_binary_vectors", data=data)
+
+    query = np.random.randint(0, 256, size=128)
+    results = tbl.search(query).to_arrow()
+    ```
```
Comment (wjones127) [inline docs/src/search.md:138]: Going forward, I think our preference is to put these tests in `python/tests/docs` and reference them with the snip operator `--8<--`. For example, these Python examples:

https://lancedb.github.io/lancedb/basic/#connect-to-a-database

Are here: 

https://github.com/lancedb/lancedb/blob/59b57055e78c723b84793688eeb108105026df37/docs/src/basic.md?plain=1#L137-L142

And the actual code is here:

https://github.com/lancedb/lancedb/blob/59b57055e78c723b84793688eeb108105026df37/python/python/tests/docs/test_basic.py#L15-L19

If we use these, all the examples are well tested.

File: rust/lancedb/src/table.rs [hunk 3]
```diff
@@ -1870,12 +1919,22 @@ impl TableInternal for NativeTable {
...
             }
 
-            let query_vector = query_vector.as_primitive::<Float32Type>();
-            scanner.nearest(
-                &column,
-                query_vector,
-                query.base.limit.unwrap_or(DEFAULT_TOP_K),
-            )?;
+            if is_binary {
+                let query_vector = arrow::compute::cast(&query_vector, &DataType::UInt8)?;
```
Comment (wjones127) [inline rust/lancedb/src/table.rs:1923]: is query_vector here already forced to be a float32? I worry about the accuracy of it here...

**Review by wjones127 — CHANGES_REQUESTED** [review]
Could you make sure to add a test that queries an IVFFlat binary index? Right now, you just create an index with it. I'm particularly worried that the casting of the query vector will mess with the accuracy.

File: docs/src/search.md (hunk 1 above)
Comment (BubbleCal) [reply docs/src/search.md]: added

File: docs/src/search.md (hunk 2 above)
Comment (BubbleCal) [reply docs/src/search.md]: added

File: python/python/lancedb/index.py [hunk 4]
```diff
@@ -355,6 +355,97 @@ class HnswSq:
     ef_construction: int = 300
 
 
+@dataclass
+class IvfFlat:
```
[the remark this answers is not in the document]
Comment (BubbleCal) [reply python/python/lancedb/index.py:359]: added

File: rust/lancedb/src/table.rs (hunk 3 above)
Comment (BubbleCal) [reply rust/lancedb/src/table.rs:1923]: yeah it's float32 here.
the values must be integers in range 0~255 (even stored as float32s) so it won't lose any accuracy?
I just added a test to verify this

**Review by wjones127 — APPROVED** [review]

**Pull request merged**